### PR TITLE
Separated cuda runtime config from build kernel and added dynamic shared config

### PR DESCRIPTION
**Pull request by WillTrojak**
The unified cache preference can now be separately set for a kernel, as well as configuring the dynamic shared memory usage and shared carveout.

### CODE REVIEW

File: pyfr/backends/cuda/blasext.py [hunk 1]
```diff
@@ -22,7 +22,9 @@ def axnpby(self, *arr, subdims=None):
 
         # Build the kernel
         kern = self._build_kernel('axnpby', src,
-                                  [np.int32]*3 + [np.intp]*nv + [dtype]*nv)
+                                  [np.int32]*3 + [np.intp]*nv + [dtype]*nv,
+                                  prefer_l1=True
```
Comment (FreddieWitherden) [inline pyfr/backends/cuda/blasext.py:26]: Can change/remove?

File: pyfr/backends/cuda/blasext.py [hunk 2]
```diff
@@ -88,6 +90,9 @@ def reduction(self, *rs, method, norm, dt_mat=None):
         # Build the reduction kernel
         rkern = self._build_kernel('reduction', src, argt)
 
+        # Configure L1
+        rkern.set_cache_pref(prefer_l1=True)
```
Comment (FreddieWitherden) [inline pyfr/backends/cuda/blasext.py:94]: As this is a reduction kernel which does use some shared memory is it worth declaring a preference?

File: pyfr/backends/cuda/driver.py [hunk 3]
```diff
@@ -198,11 +212,20 @@ def set_cache_pref(self, *, prefer_l1=None, prefer_shared=None):
         pref = self.cuda._get_cache_pref(prefer_l1, prefer_shared)
         self.cuda.lib.cuFuncSetCacheConfig(self, pref)
 
-    def exec_async(self, grid, block, stream, *args):
+    def set_shared_size(self, *, shared=0, carveout=None):
```
Comment (FreddieWitherden) [inline pyfr/backends/cuda/driver.py:215]: Include some reference to dynamic here?

File: pyfr/backends/cuda/gimmik.py [hunk 4]
```diff
@@ -35,7 +35,10 @@ def mul(self, a, b, out, alpha=1.0, beta=0.0):
 
         # Build
         fun = self._build_kernel('gimmik_mm', src,
-                                 [np.int32, np.intp]*2 + [np.int32])
+                                 [np.int32, np.intp]*2 + [np.int32]
+                                )
```
Comment (FreddieWitherden) [inline pyfr/backends/cuda/gimmik.py:39]: Can move back up

File: pyfr/backends/cuda/blasext.py [hunk 5]
```diff
@@ -22,7 +22,8 @@ def axnpby(self, *arr, subdims=None):
 
         # Build the kernel
         kern = self._build_kernel('axnpby', src,
-                                  [np.int32]*3 + [np.intp]*nv + [dtype]*nv)
+                                  [np.int32]*3 + [np.intp]*nv + [dtype]*nv
+                                 )
```
Comment (FreddieWitherden) [inline pyfr/backends/cuda/blasext.py:26]: Can move ) back up.

File: pyfr/backends/cuda/compiler.py [hunk 6]
```diff
@@ -93,7 +93,7 @@ def compile(self, name, src, flags=[]):
 
 
 class SourceModule(object):
-    def __init__(self, backend, src):
+    def __init__(self, backend, src, additional_opt=False):
```
Comment (FreddieWitherden) [inline pyfr/backends/cuda/compiler.py:96]: Still not sure about this; how about we instead expose flags to the ini file like we do with OpenMP?  Makes it nice and accessible for those who want to poke around as new features are added to NVRTC.

File: pyfr/backends/cuda/compiler.py [hunk 7]
```diff
@@ -106,18 +106,16 @@ def __init__(self, backend, src):
...
 
         # Compile to PTX
         ptx = backend.nvrtc.compile('kernel', src, flags)
 
         # Load it as a module
         self.mod = backend.cuda.load_module(ptx)
 
-    def get_function(self, name, argtypes, *, prefer_l1=None,
-                     prefer_shared=None):
+    def get_function(self, name, argtypes, *):
```
Comment (FreddieWitherden) [inline pyfr/backends/cuda/compiler.py:118]: No need for *.

File: pyfr/backends/cuda/provider.py [hunk 8]
```diff
@@ -14,10 +14,10 @@ def get_grid_for_block(block, nrow, ncol=1):
 
 class CUDAKernelProvider(BaseKernelProvider):
     @memoize
-    def _build_kernel(self, name, src, argtypes):
+    def _build_kernel(self, name, src, argtypes, **kwargs):
```
Comment (FreddieWitherden) [inline pyfr/backends/cuda/provider.py:17]: Not sure we need this?

File: doc/src/user_guide.rst [hunk 9]
```diff
@@ -128,6 +128,10 @@ Parameterises the CUDA backend with
 
     *int*
 
+5. ``cflags`` --- additional Nvidia realtime compiler (``nvrtc``) flags:
```
Comment (FreddieWitherden) [inline doc/src/user_guide.rst:131]: NVIDIA

File: pyfr/backends/cuda/gimmik.py [hunk 10]
```diff
@@ -34,8 +34,10 @@ def mul(self, a, b, out, alpha=1.0, beta=0.0):
                           alpha=alpha, beta=beta)
 
         # Build
-        fun = self._build_kernel('gimmik_mm', src,
+        fun = self._build_kernel('gimmik_mm', src,
```
Comment (FreddieWitherden) [inline pyfr/backends/cuda/gimmik.py:37]: Trailing space.

File: pyfr/backends/cuda/packing.py [hunk 11]
```diff
@@ -18,7 +18,9 @@ def pack(self, mv):
         src = self.backend.lookup.get_template('pack').render()
 
         # Build
-        kern = self._build_kernel('pack_view', src, [np.int32]*3 + [np.intp]*4)
+        kern = self._build_kernel('pack_view', src, [np.int32]*3 + [np.intp]*4,
+                                  prefer_l1=True
```
Comment (FreddieWitherden) [inline pyfr/backends/cuda/packing.py:22]: Is this still correct?

File: pyfr/backends/cuda/packing.py [hunk 12]
```diff
@@ -18,7 +18,8 @@ def pack(self, mv):
         src = self.backend.lookup.get_template('pack').render()
 
         # Build
-        kern = self._build_kernel('pack_view', src, [np.int32]*3 + [np.intp]*4)
```
Comment (FreddieWitherden) [inline pyfr/backends/cuda/packing.py:21]: No need to change this.

Comment (WillTrojak) [reply pyfr/backends/cuda/packing.py:21]: line length

Comment (FreddieWitherden) [reply pyfr/backends/cuda/packing.py:21]:
```
In [1]: len('''        kern = self._build_kernel('pack_view', src, [np.int32]*3 + [np.intp]*4)''')
Out[1]: 79
```

File: pyfr/backends/cuda/blasext.py (hunk 5 above)
Comment (FreddieWitherden) [reply pyfr/backends/cuda/blasext.py:26]: Just following up here, and should we be setting the l1 pref? (Don't think it should make any difference, but would be interested to see.)

**Comment by FreddieWitherden** [comment]
Can we do a quick profile to ensure that post-changes all of the kernels still perform the same (as some kernels have had their L1 preferences changed).

**Comment by WillTrojak** [comment]
The performance was pretty much the smae, some kernel were a little faster, some a little slower. But all were within the varaition from one profiler run to the next. Just as a point for those using Nsight compute, it seems unable to attach to pyfr when using python3.9 at the moment. So in both instantces I used python 3.8.

Also HIP has similar features, so I should probably addd these features to the HIP backend as well, but I will wait until I have access to AMD hardware to do this.

**Comment by FreddieWitherden** [comment]
If you can make the final couple of formatting changes I think we're good to merge.

**Pull request merged**